test/tools/spec-runner/utils.js

```diff
@@ -1,18 +1,42 @@
 'use strict';
 
+function getIndicesOfAuthInUrl(connectionString) {
+  const doubleSlashIndex = connectionString.indexOf('//');
+  const atIndex = connectionString.indexOf('@');
+
+  if (doubleSlashIndex === -1 || atIndex === -1) {
+    return null;
+  }
+
+  return {
+    start: doubleSlashIndex + 2,
+    end: atIndex
+  };
+}
+
+function extractAuthString(connectionString) {
+  const indices = getIndicesOfAuthInUrl(connectionString);
+  if (!indices) {
+    return null;
+  }
+
+  return connectionString.slice(indices.start, indices.end);
+}
+
 function resolveConnectionString(configuration, spec, context) {
   const isShardedEnvironment = configuration.topologyType === 'Sharded';
   const useMultipleMongoses = spec && !!spec.useMultipleMongoses;
   const username = context && context.user;
   const password = context && context.password;
-  const authSource = context && context.authSource;
+  const authSource =
+    context && context.authSource && context.authSource !== 'admin' ? context.authSource : 'admin';
```

Comment:
Because some tests specifically do pass in a value for `authSource` and if that value is not `admin`, we don't want to override it.  We just want to default to `admin`

Reply:
is that line equivalent to `context?.authSource || 'admin'`?

Reply:
Yeah, you could also use an `||`.  I can make that change.  

Unfortunately we can't use optional chaining in our raw `.js` tests because we still test on Node v12 and optional chaining wasn't added until v13

Reply:
ah yes, good call out on the chaining